x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StGeotile.java

```diff
@@ -0,0 +1,292 @@
+        }
+    }
+
+    @Evaluator(extraName = "FromFieldAndLiteral", warnExceptions = { IllegalArgumentException.class })
+    static void fromFieldAndLiteral(LongBlock.Builder results, int p, BytesRefBlock wkbBlock, @Fixed int precision) {
+        fromWKB(results, p, wkbBlock, precision, unboundedGrid);
+    }
+
+    @Evaluator(extraName = "FromFieldDocValuesAndLiteral", warnExceptions = { IllegalArgumentException.class })
+    static void fromFieldDocValuesAndLiteral(LongBlock.Builder results, int p, LongBlock encoded, @Fixed int precision) {
+        fromEncodedLong(results, p, encoded, precision, unboundedGrid);
+    }
+
+    @Evaluator(extraName = "FromFieldAndField", warnExceptions = { IllegalArgumentException.class })
+    static void fromFieldAndField(LongBlock.Builder results, int p, BytesRefBlock in, int precision) {
+        fromWKB(results, p, in, checkPrecisionRange(precision), unboundedGrid);
+    }
+
+    @Evaluator(extraName = "FromFieldDocValuesAndField", warnExceptions = { IllegalArgumentException.class })
+    static void fromFieldDocValuesAndField(LongBlock.Builder results, int p, LongBlock encoded, int precision) {
+        fromEncodedLong(results, p, encoded, checkPrecisionRange(precision), unboundedGrid);
+    }
+
+    @Evaluator(extraName = "FromLiteralAndField", warnExceptions = { IllegalArgumentException.class })
+    static long fromLiteralAndField(@Fixed BytesRef in, int precision) {
+        return unboundedGrid.calculateGridId(GEO.wkbAsPoint(in), precision);
+    }
+
+    @Evaluator(extraName = "FromFieldAndLiteralAndLiteral", warnExceptions = { IllegalArgumentException.class })
+    static void fromFieldAndLiteralAndLiteral(LongBlock.Builder results, int p, BytesRefBlock in, @Fixed GeoTileBoundedGrid bounds) {
+        fromWKB(results, p, in, bounds);
+    }
+
+    @Evaluator(extraName = "FromFieldDocValuesAndLiteralAndLiteral", warnExceptions = { IllegalArgumentException.class })
+    static void fromFieldDocValuesAndLiteralAndLiteral(
+        LongBlock.Builder results,
+        int p,
+        LongBlock encoded,
+        @Fixed GeoTileBoundedGrid bounds
+    ) {
+        fromEncodedLong(results, p, encoded, bounds);
+    }
+
+    @Evaluator(extraName = "FromFieldAndFieldAndLiteral", warnExceptions = { IllegalArgumentException.class })
+    static void fromFieldAndFieldAndLiteral(LongBlock.Builder results, int p, BytesRefBlock in, int precision, @Fixed GeoBoundingBox bbox) {
+        GeoTileBoundedGrid bounds = new GeoTileBoundedGrid(precision, bbox);
+        fromWKB(results, p, in, bounds);
+    }
+
+    @Evaluator(extraName = "FromFieldDocValuesAndFieldAndLiteral", warnExceptions = { IllegalArgumentException.class })
+    static void fromFieldDocValuesAndFieldAndLiteral(
+        LongBlock.Builder results,
+        int p,
+        LongBlock encoded,
+        int precision,
+        @Fixed GeoBoundingBox bbox
+    ) {
+        GeoTileBoundedGrid bounds = new GeoTileBoundedGrid(precision, bbox);
```

Comment:
Are we creating a GeoTileBoundedGrid per document here?

Reply:
This particular code path is for when the precision comes from the document, so yes, we are creating the predicate per document. I tried to make sure that all cases where we know stuff earlier, we create the predicate earlier. I noticed that not all predicates actually use the precision internally, but GeoTileBoundedPredicate does.

For literal precision, see line 166 where the GeoTileBoundedGrid is created for the entire execution.

Reply:
>This particular code path is for when the precision comes from the document,

Reply:
what's the use case this is trying to cover?